src/utils/saml.ts

```diff
@@ -1118,395 +81,542 @@ export const getPreValidateResponse = (
-        // Here we collect data when an IDP sends an unsigned Response.
-        // If all IDPs sign it, we can safely request it as mandatory @ref https://www.pivotaltracker.com/story/show/174710289.
-        if (eventHandler && signatureOfResponseCount === 0) {
-          eventHandler({
+      ),
+      TE.map(() => _)
+    );
+
+  const stepTwelveLeft = (error: Error | ITransformValidation): void => {
+    if (eventHandler) {
+      TransformError.is(error)
+        ? eventHandler({
             data: {
-              idpIssuer: _.SAMLRequestCache.idpIssuer,
-              message: "Missing Request signature"
+              idpIssuer: error.idpIssuer,
+              message: error.message,
+              numberOfTransforms: String(error.numberOfTransforms)
             },
-            name: "spid.error.signature",
-            type: "INFO"
+            name: "spid.error.transformOccurenceOverflow",
+            type: "ERROR"
+          })
+        : eventHandler({
+            data: {
+              message: error.message
+            },
+            name: "spid.error.generic",
+            type: "ERROR"
           });
-        }
-        return callback(null, true, _.InResponseTo);
-      }
+    }
+    return callback(E.toError(error.message));
+  };
+
+  const stepTwelveRight = (_: IStepEleven): void => {
+    // Number of the Response signature.
+    // Calculated as number of the Signature elements inside the document minus number of the Signature element of the Assertion.
+    const signatureOfResponseCount =
+      _.Response.getElementsByTagNameNS(SAML_NAMESPACE.XMLDSIG, "Signature")
+        .length -
+      _.Assertion.getElementsByTagNameNS(SAML_NAMESPACE.XMLDSIG, "Signature")
+        .length;
+    // For security reasons it is preferable that the Response be signed.
+    // According to the technical rules of SPID, the signature of the Response is optional @ref https://docs.italia.it/italia/spid/spid-regole-tecniche/it/stabile/single-sign-on.html#response.
+    // Here we collect data when an IDP sends an unsigned Response.
+    // If all IDPs sign it, we can safely request it as mandatory @ref https://www.pivotaltracker.com/story/show/174710289.
+    if (eventHandler && signatureOfResponseCount === 0) {
+      eventHandler({
+        data: {
+          idpIssuer: _.SAMLRequestCache.idpIssuer,
+          message: "Missing Request signature"
+        },
+        name: "spid.error.signature",
+        type: "INFO"
+      });
+    }
+    return callback(null, true, _.InResponseTo);
+  };
+
+  return pipe(
+    stepOne,
+    TE.chain(stepTwo),
+    TE.chain(stepThree),
+    TE.chain(stepFour),
+    TE.chain(stepFive),
+    TE.chain(stepSix),
+    TE.chain(stepSeven),
+    TE.chain(stepEight),
+    TE.chain(stepNine),
+    TE.chain(stepTen),
+    // check for Transform over SAML Response
+    TE.chainW(stepEleven),
+    TE.bimap(
+      error => stepTwelveLeft(error),
+      _ => stepTwelveRight(_)
     )
```

Comment:
Great work in here. I like a lot the splitting of the giant `TE` chain. Now we can improve code clearness using semantic names for each validation step. Below an example, but you are free to use another naming convention.
The same thing must be do with the types use in input and output for all steps.
```suggestion
  return pipe(
    responseElementValidationStep,
    TE.chain(returnRequestAndResponseStep),
    TE.chain(parseSAMLRequestStep),
    TE.chain(getIssueInstantFromRequestStep),
    TE.chain(issueInstantValidationStep),
    TE.chain(assertionIssueInstantValidationStep),
    TE.chain(authnContextClassRefValidationStep),
    TE.chain(attributesValidationStep),
    TE.chain(responseIssuerValidationStep),
    TE.chain(assertionIssuerValidationStep),
    TE.chainW(transformValidationStep),
    TE.bimap(validationFailure, validationSuccess)
```

Reply:
I lazly enumerated them, now is a lot better! thank you!